### ISSUE TITLE
golangci-lint, script wrapped go binary, and malformed import path errors

### ISSUE THREAD

**Issue by BenjamenMeyer** [golangci/golangci-lint#6173]
### Welcome

- [x] Yes, I'm using a binary release within 2 latest [releases](https://github.com/golangci/golangci-lint/releases). Only such installations are supported.
- [x] Yes, I've searched similar [issues on GitHub](https://github.com/golangci/golangci-lint/issues) and didn't find any.
- [x] Yes, I've read the `typecheck` section of the [FAQ](https://golangci-lint.run/docs/welcome/faq/#why-do-you-have-typecheck-errors).
- [x] Yes, I've tried with the standalone [linter](https://golangci-lint.run/docs/linters/) if available (e.g., gocritic, go vet, etc.).
- [x] I agree to follow this project's [Code of Conduct](https://github.com/golangci/golangci-lint?tab=coc-ov-file)

### How did you install golangci-lint?

Official binary

### Description of the problem

When running `golangci-lint run` I get errors like the following:
```
ERROR Running error: context loading failed: failed to load packages: failed to load packages: failed to load with go/packages: err: exit status 1: stderr: malformed import path "{{context.Compiler}}": invalid char '{'
```

I've seen this for a long time across numerous versions of Go and golangci-lint, and searching always turned up issues like https://github.com/golangci/golangci-lint/issues/4659 which suggested it was an internal Golang issue.

Recently though I found that the issue can be easily reproduced by simply using a wrapper script for calling the Go binary such as the following:
```bash
#!/bin/bash
go1.25.1 $@
```

Background: I typically have had a script effectively similar to this so that I can be using the latest version of Go I have installed when simply calling `go`; the script uses Bash functions to simply look at the files in ~/go/bin and determine which version of Go is the latest. However the above is sufficient to reproduce the issue.

I'm curious why this simple script would cause such a problem.

### Version of golangci-lint

<details>

```console
$ golangci-lint --version
golangci-lint has version 2.6.1 built with go1.25.3 from e3b3bac5 on 2025-11-04T11:05:57Z
```

</details>

### Configuration

Some of the linters do not seem to have an issue; while others do.

<details>

The configuration below contains a series of linters that each individually display the issue. However, running tools like `go vet` directly do not have this issue; only running them through `golangci-lint`.

```yml
linters:
  default: none
  enable:
    - errcheck
    - goconst
    - gocritic
    - gosec
    - govet
    - unused
  settings:
    errcheck:
      check-type-assertions: true
    goconst:
      min-len: 2
      min-occurences: 3
    gocritic:
      enable-tags:
        - style
formatters:
  enable:
    - gofmt
    - goimports
version: "2"
run:
  issues-exit-code: 1
```

</details>

### Go environment

<details>

```console
$ go version && go env
go version go1.25.3 linux/amd64
AR='ar'
CC='gcc'
CGO_CFLAGS='-O2 -g'
CGO_CPPFLAGS=''
CGO_CXXFLAGS='-O2 -g'
CGO_ENABLED='1'
CGO_FFLAGS='-O2 -g'
CGO_LDFLAGS='-O2 -g'
CXX='g++'
GCCGO='gccgo'
GO111MODULE=''
GOAMD64='v1'
GOARCH='amd64'
GOAUTH='netrc'
GOBIN=''
GOCACHE='/home/bmeyer/.cache/go-build'
GOCACHEPROG=''
GODEBUG=''
GOENV='/home/bmeyer/.config/go/env'
GOEXE=''
GOEXPERIMENT=''
GOFIPS140='off'
GOFLAGS=''
GOGCCFLAGS='-fPIC -m64 -pthread -Wl,--no-gc-sections -fmessage-length=0 -ffile-prefix-map=/tmp/go-build3711301791=/tmp/go-build -gno-record-gcc-switches'
GOHOSTARCH='amd64'
GOHOSTOS='linux'
GOINSECURE=''
GOMOD='/dev/null'
GOMODCACHE='/home/bmeyer/go/pkg/mod'
GONOPROXY=''
GONOSUMDB=''
GOOS='linux'
GOPATH='/home/bmeyer/go'
GOPRIVATE=''
GOPROXY='https://proxy.golang.org,direct'
GOROOT='/home/bmeyer/sdk/go1.25.3'
GOSUMDB='sum.golang.org'
GOTELEMETRY='local'
GOTELEMETRYDIR='/home/bmeyer/.config/go/telemetry'
GOTMPDIR=''
GOTOOLCHAIN='auto'
GOTOOLDIR='/home/bmeyer/sdk/go1.25.3/pkg/tool/linux_amd64'
GOVCS=''
GOVERSION='go1.25.3'
GOWORK=''
PKG_CONFIG='pkg-config'
```

</details>

### Verbose output of running

<details>

```console
$ golangci-lint cache clean
$ golangci-lint run -v
INFO golangci-lint has version 2.6.1 built with go1.25.3 from e3b3bac5 on 2025-11-04T11:05:57Z 
INFO [config_reader] Config search paths: [./ /home/bmeyer/Devel/personal/golintci-problem /home/bmeyer/Devel/personal /home/bmeyer/Devel /home/bmeyer /home /] 
INFO maxprocs: Leaving GOMAXPROCS=16: CPU quota undefined 
INFO [goenv] Read go env for 3.783163ms: map[string]string{"GOCACHE":"/home/bmeyer/.cache/go-build", "GOROOT":"/usr/lib/go-1.24"} 
INFO [lintersdb] Active 5 linters: [errcheck govet ineffassign staticcheck unused] 
INFO [loader] Go packages loading at mode 8767 (types_sizes|compiled_files|exports_file|imports|deps|files|name) took 9.067034ms 
ERRO Running error: context loading failed: failed to load packages: failed to load packages: failed to load with go/packages: err: exit status 1: stderr: malformed import path "{{context.Compiler}}": invalid char '{'
malformed import path "--": leading dash 
INFO Memory: 2 samples, avg is 40.3MB, max is 40.3MB 
INFO Execution took 13.080118ms
```

</details>

### A minimal reproducible example or link to a public repository

A user may for a variety of reasons want to wrap a call to `go` with a script. In my personal case I do it to manage multiple installations of `go` and be able to easily upgrade the default version. However, this seems to cause an issue for `golangci-lint` which results in an error that is not easily diagnosed, and searching online has not yielded any good results.

Now a workaround is to use symlinks instead; however they are not quite as flexible as a script since the script could auto-detect newer versions to auto-update the default go binary pointed to by go where a symlink needs to be manually updated.

NOTE: I also wanted to document this for others to find since over the last 6 years that I've seen this issue I have yet to see anyone point to any kind of resolution or potential real cause. Hopefully by at least documenting this it'll help others as well.

<details>

Sample program `main.go` (basic hello world)

```go
package main

import "fmt"

func main() {
        fmt.Println("Hello World!")
}
```
Sample `go.mod`:
```go
module main

go 1.25.3
```

Sample run:

```bash
$ golangci-lint run
ERRO Running error: context loading failed: failed to load packages: failed to load packages: failed to load with go/packages: err: exit status 1: stderr: malformed import path "{{context.Compiler}}": invalid char '{'
malformed import path "--": leading dash 
```

Sample script to wrap the `go` binary:
```bash
#!/bin/bash

go1.25.3 $@
```

The script above should be more or less equivalent from an operational view between calling a `go` binary directly vs calling a script wrapping a call to a `go` binary. So I'm quite curious as to why that simple change would cause golangci-lint to fail like this.

</details>

### Validation

- [x] Yes, I've included all information above (version, config, etc.).

### Supporter

- [ ] I am a sponsor/backer through [GitHub](https://github.com/sponsors/golangci/) or [OpenCollective](https://opencollective.com/golangci-lint)

**Comment by ldez** [comment]
Non-namespaced module names (`module main`) are theoretically reserved for Go itself.

Those module names should not be used.

**Comment by BenjamenMeyer** [comment]
Background: I have been trying to figure out this particular problem for a few years. As a developer I tend to keep a number of Go installs around, installed via `go install`, for example:
```bash
$ go install golang.org/dl/go1.25.3@latest
$ go1.25.3 download
```
As a matter of curiosity I swapped the script for a symlink and noticed the issue went away.

The output has been the same across go version and versions of golangci-lint.

Hopefully this will help folks by at least documenting something as a possible path of a cause.

**Comment by BenjamenMeyer** [comment]
> Non-namespaced module names (`module main`) are theoretically reserved for Go itself.
> 
> Those module names should not be used.

The `go.mod` is just a sample so that `go vet` would work; it is very much *not* the cause here. Try any `module` you like, you'll get the same response.

**Comment by ldez** [comment]
> Try any module you like, you'll get the same response.

With your example, but with a namespace module name:

```
module github.com/golangci/sandbox

go 1.25.3
```

```console
$ golangci-lint version
golangci-lint has version 2.6.1 built with go1.25.3 from e3b3bac5 on 2025-11-04T11:05:57Z
$ golangci-lint run    
0 issues.
```

**Comment by BenjamenMeyer** [comment]
@ldez I updated my sample `go.mod` to use something else:

```bash
$ cat go.mod
module github.com/benjamenmeyer/helloworld

go 1.25.3
$ golangci-lint run
ERRO Running error: context loading failed: failed to load packages: failed to load packages: failed to load with go/packages: err: exit status 1: stderr: malformed import path "{{context.Compiler}}": invalid char '{'
malformed import path "--": leading dash 
```
The issue persists. Again, I've run this at work with numerous projects that had `module` paths pointing at Git Repositories, etc.

The issue seems to stem from `go` not being a binary but being a script that wraps the binary.

**Comment by ldez** [comment]
Sorry, I missed your Go alias.

golangci-lint depends on Go tooling.
The tooling calls `go list` programmatically.
The tooling is outside golangci-lint code base.

The `go build` command is able to install and use the right Go version automatically.

Why do you need to manage multiple versions of Go manually?

**Comment by ldez** [comment]
OK, I read your issue again: with your script, you are trying to always use the latest installed version of Go.

I think this is not the right approach to handle the Go version, but anyway.

As I said, the error comes from the Go tooling, to be precise, from the function [`packages.Load()`](https://github.com/golang/tools/blob/277e1b1bbd113849eb75163072c50d9dc79381b7/go/packages/packages.go#L261).
The `go` command is invoked [here](https://github.com/golang/tools/blob/277e1b1bbd113849eb75163072c50d9dc79381b7/go/packages/external.go#L121) and [here](https://github.com/golang/tools/blob/277e1b1bbd113849eb75163072c50d9dc79381b7/internal/gocommand/invoke.go#L248).

We have no control over this because this is outside the golangci-lint code.

**Comment by ldez** [comment]
I'm trying to understand the behavior, but for now, I don't have a precise explanation.

I will try to find the explanation tomorrow.

**Comment by ldez** [comment]
I spent 2 hours debugging the wrong go command 😄 

I can reproduce without golangci-lint:

```console
$ go list -f "{{context.GOARCH}} {{context.Compiler}}" -- unsafe
malformed import path "{{context.Compiler}}": invalid char '{'
malformed import path "--": leading dash

$ go1.25.3 list -f "{{context.GOARCH}} {{context.Compiler}}" -- unsafe
amd64 gc
```

**Comment by ldez** [comment]
This is because bash is removing the double quotes when using `$@`.

---

<details><summary>Note for me to reproduce the error</summary>

```bash
docker run --rm -v $(pwd):/app -w /app -it golang:1.24 bash
```

```bash
go install golang.org/dl/go1.25.3@latest
go1.25.3 download
rm /usr/local/go/bin/go
cat > /usr/local/go/bin/go  <<EOF
#!/bin/bash

# printf "%q\n" "$@"

go1.25.3 \$@

EOF
chmod +x /usr/local/go/bin/go
go version
git config --global --add safe.directory /app
```

```bash
apt update && apt -y upgrade
apt install -y vim

vim /usr/local/go/bin/go
```

</details>

**Comment by BenjamenMeyer** [comment]
> This is because bash is removing the double quotes when using `$@`.

Interesting... and thank you very much for the diagnosis. Hopefully this will help point others seeing this issue as well.
I can confirm that quoting the `$@` resolves my local reproduction of the issue.
